interfaces/builtin/common_files.go

```diff
@@ -0,0 +1,162 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/snap"
+)
+
+type commonFilesInterface struct {
+	commonInterface
+
+	apparmorHeader    string
+	extraPathValidate func(string) error
+}
+
+// filesAAPerm can either be files{Read,Write} and converted to a string
+// expands into the right apparmor permissions for the files interface.
+type filesAAPerm int
+
+const (
+	filesRead filesAAPerm = iota
+	filesWrite
+)
+
+func (a filesAAPerm) String() string {
+	switch a {
+	case filesRead:
+		return "rk,"
+	case filesWrite:
+		return "rwkl,"
+	}
+	panic(fmt.Sprintf("invalid perm: %d", a))
+}
+
+func formatPath(ip interface{}) (string, error) {
+	p, ok := ip.(string)
+	if !ok {
+		return "", fmt.Errorf("%[1]v (%[1]T) is not a string", ip)
+	}
+	prefix := ""
+	if strings.HasPrefix(p, "$HOME") && strings.Count(p, "$HOME") == 1 {
+		p = strings.Replace(p, "$HOME", "@{HOME}", 1)
+		prefix = "owner "
+	}
+	p += "{,/,/**}"
+
+	return fmt.Sprintf("%s%q", prefix, filepath.Clean(p)), nil
+}
+
+func allowPathAccess(buf *bytes.Buffer, perm filesAAPerm, paths []interface{}) error {
+	for _, rawPath := range paths {
+		p, err := formatPath(rawPath)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(buf, "%s %s\n", p, perm)
+	}
+	return nil
+}
+
+func (iface *commonFilesInterface) validatePaths(attrName string, paths []interface{}) error {
```

Comment:
Not a strong opinion or something to change now but I think that having a set of things like validate one path, validate a list of paths, etc, would be useful as plain functions. I suspect that more than one interface needs to have code like this.

Reply:
I would wait to break this out until there is a need to-- these are needed for quite specific reasons atm.